src/CallbackStream.php

```diff
@@ -0,0 +1,182 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @see       http://github.com/zendframework/zend-diactoros for the canonical source repository
+ * @copyright Copyright (c) 2015 Oscar Otero (http://oscarotero.com) / Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-diactoros/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Diactoros;
+
+use InvalidArgumentException;
+use RuntimeException;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Implementation of PSR HTTP streams
+ */
+class CallbackStream implements StreamInterface
+{
+    /**
+     * @var callable|null
+     */
+    protected $callback;
+
+    /**
+     * @param callable $callback
+     * @throws InvalidArgumentException
+     */
+    public function __construct(callable $callback)
+    {
+        $this->attach($callback);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->getContents();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        $this->callback = null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function detach()
+    {
+        $callback = $this->callback;
+        $this->callback = null;
+        return $callback;
+    }
+
+    /**
+     * Attach a new callback to the instance.
+     *
+     * @param callable $callback
+     * @throws InvalidArgumentException for callable callback
+     */
+    public function attach(callable $callback)
+    {
+        $this->callback = $callback;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSize()
+    {
+    }
```

Comment:
This is specified as MUST return an integer. Considering we just added support to the `SapiEmitter` to auto-emit a Content-Length header based on the return value of this method, it's even more critical.

This would likely require that you return the strlen() of the return of getContents() — which would mean you'll likely want to have that method cache the results of calling the callback.

Doing so would also allow you to do things like read(), tell(), etc, as these could call `getContents()` if the content is not yet cached, and then start operating on the cached value. That would also allow better flexibility with emitters, as they would not need to typehint against this stream type to determine how to send the stream.